design/fips-compliance.md

```diff
@@ -0,0 +1,213 @@
+versions will bring to the development team.
+
+### Red Hat image
+
+We will also release a FIPS compliant Red Hat image of the `conjur-authn-k8s
+-client` to our Red Hat registry. Before we do that, we should create a 
+`conjur-authn-k8s-client` FIPS project in the RH container registry.
+
+TBD - how do we create the FIPS project in the RH container registry?
+
+### Performance
+
+Same SLA of performance should be kept. As of today, we don't have any
+performance tests and an SLA (that I know of). We should decide on one and
+verify that the change to GoBoring doesn't impact the performance.
+
+We should write a performance test and run it twice - once on the `master
+` image and once on the new FIPS-compliance image - and verify that the
+ performance is not affected by the change. 
+
+## Test plan
+
+Before we dig into which tests should run, we should decide where, and how, they
+should run. Currently, the `conjur-authn-k8s-client` have UTs in its project
+and a vanilla test that runs in `kubernetes-conjur-demo`. The `kubernetes-conjur-demo`
+is triggered only when we push code to that repository. 
+
+Even while putting aside the fact that we have only a vanilla test, our current
+ test flow is still not optimal. 
+ Let's propose some options to improve our flow, and decide on the best one.
+
+### Trigger `kubernetes-conjur-demo` to run after every build of `conjur-authn-k8s-client`
+
+We can find a way to run `kubernetes-conjur-demo` using the currently built
+ authenticator-client, utilizing the `LOCAL_AUTHENTICATOR` environment variable.
+ If this is possible, we can trigger a `kubernetes-conjur-demo` build as part
+  of our build.
+
+This will give us full confidence in the green build as it also passed
+ integration tests.
+ 
+ However, the build time of `kubernetes-conjur-demo` is ~25 minutes (opposing
+  2-3 minutes of the `conjur-authn-k8s-client`) and it is not very stable
+  . This means that it will harder to merge PRs into the `conjur-authn-k8s
+  -client`.
+  
+I am not sure that this can be implemented. By setting`LOCAL_AUTHENTICATOR
+` to `true` we use the last built image (ran in the `./bin/build` step. In
+ case we have builds running in parallel it may affect the latest built image
+  so it is possible that we can't use a local authenticator in Jenkins.
+
+### Trigger `kubernetes-conjur-demo` to run after every master build of `conjur-authn-k8s-client`
+
+This option will still not fail `conjur-authn-k8s-client` builds that will
+fail in `kubernetes-conjur-demo` but we will get _some_ feedback. 
+After we will merge the PR into `master` we can follow the `master` build of 
+`kubernetes-conjur-demo` to see if it passed. This is better than a nightly
+build also because at the end of the day we will see which PR
+introduced a failure and it will be easier to fix the error. 
+  
+### Add integration tests to `conjur-authn-k8s-client` as we have in the `secrets-provider-for-kubernetes`
```

Comment:
I'm a fan of this option and we could use KinD (see previous link) and GitHub actions to make this runnable for contributors too.

Reply:
@sgnn7 those tests include preparing the environment and deploying conjur (take ~8 min ) and tests (take ~5 min and if we will add some performance tests it will increase ).
I dont think that running all tests on each commit is the right way, maybe doing it as the last step before merge ?

Reply:
> doing it as the last step before merge ?

Yes, this is what I meant. Any master merge and any pull request should have this running but other cases probably don't need to

Reply:
i don't understand. what are the cases other than before merging to master? why would you run a build that is not merged to master?

do you mean that build of master will not run the tests?

Reply:
@sgnn7 @eladkug please comment on the above :)

Reply:
I think that only master need to run full tests + when merging to master, during the development i dont think we need to run full tests (maybe basic vanilla test)

Reply:
By running tests before merging (also known as "gated check-ins"), you can be sure that code works before it hits master (which should be always in working order). If you don't, the chance that we will be able to cut a release whenever we want is greatly reduced (and broken merges may interrupt other developers as they hit the same bug on rebases) since a merge is only assumed to be working until it lands into master that would confirm it. [This thread](https://softwareengineering.stackexchange.com/questions/260778/is-it-a-good-practice-to-run-unit-tests-in-version-control-hooks) has some different points of view on this but keep in mind that the accepted solution on it is not the highest-voted comment.